Invoices.elm

```diff
@@ -75,52 +81,31 @@ type Msg
 
 
 update : Msg -> Model -> Model
-update msg model =
+update msg ({ products } as model) =
     case msg of
         AddProduct ->
-            { model | products = model.products ++ [ model.products |> List.length |> createProduct ] }
+            addNew model
 
         ChangeUnit id newUnit ->
             let
-                changeUnit product =
-                    if product.id == id then
-                        { product
-                            | unit = Result.withDefault product.unit <| String.toInt newUnit
-                        }
-                    else
-                        product
+                updateUnit product =
+                    { product | unit = Result.withDefault product.unit <| String.toInt newUnit }
             in
-                { model
-                    | products = List.map changeUnit model.products
-                }
+                updateProducts (updateWithId id updateUnit) model
 
         ChangeTax id newTax ->
             let
-                changeTax product =
-                    if product.id == id then
-                        { product
-                            | tax = Result.withDefault product.tax <| String.toFloat newTax
-                        }
-                    else
-                        product
+                updateTax product =
+                    { product | tax = Result.withDefault product.tax <| String.toFloat newTax }
             in
-                { model
-                    | products = List.map changeTax model.products
-                }
+                updateProducts (updateWithId id updateTax) model
 
         ChangePrice id newPrice ->
             let
-                changePrice product =
-                    if product.id == id then
-                        { product
-                            | price = Result.withDefault product.price <| String.toFloat newPrice
-                        }
-                    else
-                        product
+                updatePrice product =
+                    { product | price = Result.withDefault product.price <| String.toFloat newPrice }
             in
-                { model
-                    | products = List.map changePrice model.products
-                }
+                updateProducts (updateWithId id updatePrice) model
```

Comment:
most of the repeating patterns were removed. It would be simple enough to go even further (see the description of PR) however I would say going too far would be definitely harmful to the extensibility of this app.